nmosnode/aggregator.py

```diff
@@ -83,58 +80,288 @@ def __init__(self, logger=None, mdns_updater=None):
+        If a registration fails then another aggregator will be tried."""
+        self.logger.writeDebug("Entering Discovery Mode")
+        # Wait backoff period
+        if not self._aggregator_failure:
+            self._back_off_timer()
+
+        self._aggregator_failure = False
+
+        # Update cached list of aggregators
+        if self._aggregator_list_stale:
+            self._flush_cached_aggregators()
+        while True:
+            self.aggregator = self._get_aggregator()
+            if self.aggregator is None:
+                self.logger.writeDebug("Failed to find registration API")
+                break
+            self.logger.writeDebug("Aggregator set to: {}".format(self.aggregator))
+
+            if self._heartbeat_operation():
+                # Successfully registered Node
+                break
+
+    def _registered_operation(self):
+        """In Registered operation, the Node is registered so a heartbeat will be performed,
+        if the heartbeat is successful the Node will wait 5 seconds before attempting another heartbeat.
+        Else another aggregator will be selected"""
+        if not self._heartbeat_operation():
+            # Heartbeat failed
+            # Flag to update cached list of aggregators
+            self.aggregator = None
+            self._aggregator_failure = True
+
+    def _heartbeat_operation(self):
+        """Performs a heartbeat to registered aggregator
+        If heartbeat fails it will take actions to correct the error, by re-registering the Node
+        If successfull will return True, else will return False"""
+        if not self.aggregator:
+            return False
+        try:
+            self.logger.writeDebug("Sending heartbeat for Node {}"
+                                   .format(self._node_data["node"]["data"]["id"]))
+            R = self._send("POST", self.aggregator, self.aggregator_apiversion,
+                           "/health/nodes/{}".format(self._node_data["node"]["data"]["id"]))
+
+            if R.status_code == 200 and self._node_data["registered"]:
+                # Continue to registered operation
+                self._registered()
+                heartbeat_wait = 5
+                while heartbeat_wait > 0 and self._running:
+                    gevent.sleep(1)
+                    heartbeat_wait -= 1
+                return True
+
+            elif R.status_code in [200, 409]:
+                # Delete node from registry
+                if self._unregister_node(R.headers.get('Location')):
+                    return self._register_node()
+                else:
+                    # Try next registry
+                    return False
+
+        except InvalidRequest as e:
+            if e.status_code == 404:
+                # Re-register
+                self.logger.writeWarning("404 error on heartbeat. Marking Node for re-registration")
+                self._node_data["registered"] = False
+                return self._register_node()
```

Comment:
This method can throw an exception which doesn't appear to be caught anywhere. The state diagram probably expects the equivalent of a return False here if this happens.

Reply:
All exceptions in register node should be caught, it should only return True or False